File: scripts/cleanup_duplicates.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Find;

# Headers that represent the new high-value action sections
my @action_headers = (
    "## Investor Action",
    "## Walking Action",
    "## Family Action",
    "## Practical Action",
    "## Tour Action"
);

# Headers containing these strings should be removed entirely
my @remove_patterns = (
    "Checklist",
    "체크리스트",
    "チェックリスト",
    "철학적 리플렉션"
);

sub process_file {
    my $file = $_;
    return unless $file =~ /\.md$/;
    
    my $filepath = $File::Find::name;
    
    open my $fh, '<', $file or die "Could not open $file: $!";
    my $content = do { local $/; <$fh> };
    close $fh;
    
    my $modified = 0;
    
    # 1. Remove sections with specific removal patterns
    foreach my $pattern_str (@remove_patterns) {
        my $pattern = qr/(?m)^## .*?\Q$pattern_str\E.*?\n.*?(?=\n## |\Z)/s;
        if ($content =~ s/$pattern//g) {
            print "Removed redundant section with pattern '$pattern_str' in $filepath\n";
            $modified = 1;
        }
    }
    
    # 2. Deduplicate Action sections
    foreach my $header (@action_headers) {
        # Match header line + content until next header or end
        my $pattern = qr/(?m)^\Q$header\E:.*?\n.*?(?=\n## |\Z)/s;
        my @matches = $content =~ /$pattern/g;
        
        if (scalar @matches > 1) {
            print "Found duplicates of $header in $filepath\n";
            # Keep only the LAST match (often the one added later is the most up-to-date in this specific workflow)
            my $last = $matches[-1];
            
            my @parts = split(/$pattern/, $content);
            # We want to keep $parts[0], then $last, then join the rest of $parts
            # Actually, the parts are the bits *between* matches.
            # So if we have 3 matches, we have 4 parts.
            # parts[0] match[0] parts[1] match[1] parts[2] match[2] parts[3]
            # We want to keep only one match.
            
            # Let's just find the first match's position and put the match there.
            my $new_content = $parts[0] . $last;
            for (my $i = 1; $i < @parts; $i++) {
                $new_content .= $parts[$i];
            }
            $content = $new_content;
            $modified = 1;
        }
    }
    
    if ($modified) {
        # Normalize spacing
        $content =~ s/\n{3,}/\n\n/g;
        
        open my $wh, '>', $file or die "Could not write to $file: $!";
        print $wh $content;
        close $wh;
        print "Processed $filepath\n";
    }
}

find(\&process_file, "src/data/blog");

File: scripts/cleanup_duplicates.py
import os
import re

def cleanup_duplicates(directory):
    action_headers = [
        "## Investor Action",
        "## Walking Action",
        "## Family Action",
        "## Practical Action"
    ]
    
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith(".md"):
                filepath = os.path.join(root, file)
                with open(filepath, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                modified = False
                new_content = content
                
                for header in action_headers:
                    # Find all occurrences of the section starting with the header
                    # and ending before the next header (## ) or end of file
                    pattern = rf"(?m)^{re.escape(header)}.*?(?=\n## |\Z)"
                    sections = re.findall(pattern, new_content, re.DOTALL)
                    
                    if len(sections) > 1:
                        print(f"Found {len(sections)} duplicates of '{header}' in {filepath}")
                        # Keep only the first occurrence
                        first_occurrence = sections[0].strip()
                        
                        # Replace all occurrences with just the first one
                        # We use a trick to replace the whole sequence of matches
                        # but that might be tricky if they are scattered.
                        # However, they seem to be contiguous in the files I saw.
                        
                        # Let's try to remove all occurrences and then re-insert one at the location of the first one.
                        # Actually, a simpler way: just replace the redundant ones with empty string.
                        
                        # To avoid replacing parts of other text, we'll replace the full matched sections.
                        # But since they might be identical, we can't just use .replace() easily for 'all but one'.
                        
                        # Refined approach: split by the pattern and join back with only one instance.
                        parts = re.split(pattern, new_content, flags=re.DOTALL)
                        # parts will have (len(sections) + 1) elements.
                        # We want: part[0] + sections[0] + part[1] + (empty) + part[2] ...
                        
                        # Reconstruct:
                        reconstructed = parts[0] + sections[0].strip() + "\n\n"
                        for i in range(1, len(parts)):
                            # Only add the part if it's not just whitespace between duplicates
                            part_content = parts[i].strip()
                            if part_content:
                                reconstructed += parts[i]
                            elif i < len(parts) - 1:
                                # Skip whitespace between duplicates
                                pass
                        
                        new_content = reconstructed
                        modified = True
                
                if modified:
                    with open(filepath, 'w', encoding='utf-8') as f:
                        f.write(new_content.strip() + "\n")
                    print(f"Cleaned up {filepath}")

if __name__ == "__main__":
    blog_dir = "src/data/blog"
    cleanup_duplicates(blog_dir)
